AWS: remove volume IOPS limit

jacobmarble opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Max IOPS for SSD (io1) volumes was increased from 20,000 to 32,000:
https://aws.amazon.com/about-aws/whats-new/2017/12/amazon-ebs-provisioned-iops-ssd--io1--volumes-now-support-32-000-iops-and-500-mbs-per-volume/

and later to 64,000:
https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-elastic-block-store-announces-double-the-performance-of-provisioned-iops-volumes/

Rather than chasing this limit, allow the AWS SDK to surface errors for IOPS value out of range.

**Which issue(s) this PR fixes**:

(none found)

**Special notes for your reviewer**:

I'm proposing the same for CSI:
https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/483

**Does this PR introduce a user-facing change?**:

```release-note
Increased maximum IOPS of AWS EBS io1 volumes to 64,000 (current AWS maximum).
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html

jacobmarble commented:
I have signed the CLA. Let's see if this message triggers a recheck.

jacobmarble commented:
I have manually tested this change. A PV/PVC based on `io1` with <64,000 IOPS created successfully. A PV/PVC based on `io1` with >64,000 IOPS did not. The errors exposed in logs and `kubectl described` are short, clear, useful.

Resource config:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: data-io1-1000g-50000iops
  namespace: default
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1000G
  storageClassName: io1-50
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: data-io1-1500g-75000iops
  namespace: default
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1500G
  storageClassName: io1-50
---
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: io1-50
parameters:
  iopsPerGB: "50"
  type: io1
provisioner: kubernetes.io/aws-ebs
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

Results of `kubectl describe` on failing PVC:
```
Name:          data-io1-1500g-75000iops
Namespace:     default
StorageClass:  io1-50
Status:        Pending
Volume:
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/aws-ebs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
VolumeMode:    Filesystem
Mounted By:    <none>
Events:
  Type     Reason              Age   From                         Message
  ----     ------              ----  ----                         -------
  Warning  ProvisioningFailed  89s   persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: 20c19f61-7ceb-4275-baf3-a51dd9dd067d
  Warning  ProvisioningFailed  86s  persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: 2abbb3e3-ffa3-4815-bfb1-cd4c410455c2
  Warning  ProvisioningFailed  71s  persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: 666f9642-e454-4c3d-b9a9-f347223ce033
  Warning  ProvisioningFailed  56s  persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: c7bcee49-d5c4-4015-b32a-c5fd409fcfba
  Warning  ProvisioningFailed  41s  persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: e0e5d299-014c-4c60-ab50-524283b24679
  Warning  ProvisioningFailed  26s  persistentvolume-controller  Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
           status code: 400, request id: b1bf92e1-76b1-4a7e-a589-418d470f1b97
```

Kubernetes log:
```
I0430 05:58:31.041343   15843 aws_util.go:110] Error creating EBS Disk volume: InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
	status code: 400, request id: 6ccfc493-5141-454d-a578-04e659708067
E0430 05:58:31.041377   15843 aws_ebs.go:494] Provision failed: InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
	status code: 400, request id: 6ccfc493-5141-454d-a578-04e659708067
I0430 05:58:31.041415   15843 pv_controller.go:1499] failed to provision volume for claim "default/data-io1-1500g-75000iops" with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
	status code: 400, request id: 6ccfc493-5141-454d-a578-04e659708067
E0430 05:58:31.041467   15843 goroutinemap.go:150] Operation for "provision-default/data-io1-1500g-75000iops[5bab79c7-7517-4a15-a6af-485d2ceb55e7]" failed. No retries permitted until 2020-04-30 05:59:35.041448037 +0000 UTC m=+3039.192762601 (durationBeforeRetry 1m4s). Error: "InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.\n\tstatus code: 400, request id: 6ccfc493-5141-454d-a578-04e659708067"
I0430 05:58:31.041528   15843 event.go:278] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"data-io1-1500g-75000iops", UID:"5bab79c7-7517-4a15-a6af-485d2ceb55e7", APIVersion:"v1", ResourceVersion:"934", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' Failed to provision volume with StorageClass "io1-50": InvalidParameterValue: Volume iops of 69850 is too high; maximum is 64000.
	status code: 400, request id: 6ccfc493-5141-454d-a578-04e659708067
I0430 05:58:31.232585   15843 pvc_protection_controller.go:263] PVC default/data-io1-1500g-75000iops is unused
I0430 05:58:31.235708   15843 pvc_protection_controller.go:209] Removed protection finalizer from PVC default/data-io1-1500g-75000iops
```

@gnufied @justinsb please share your thoughts.

leakingtapan commented:
/ok-to-test

leakingtapan commented:
/lgtm

jacobmarble commented:
Matching PR for AWS/EBS CSI driver [has also been tested](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/483#issuecomment-646347334).

Please advise next steps.

jacobmarble commented:
Friendly note here. I believe that this and https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/483 are ready for consideration to be merged.

leakingtapan commented:
/lgtm
/approve

@gnufied mind approve as well?

nikhita commented:
/assign @jsafrane 
for approval

jsafrane commented:
I'm sorry I missed this one. Still, AWS EBS storage class offers "iops per GB". Users should not get an error when they want too big volume or too small volume. IMO it's still better to cap the value at the current maximum.

jacobmarble commented:
@jsafrane please re-read the original comment on this PR. The current AWS limit is 64,000. That number has changed twice in recent history and will likely change again in the future.

Please reconsider. I'm a user and this is the correct behavior from my perspective.

jsafrane commented:
Maybe I did not express clearly, "cap the value at the current maximum" means to bump the limit to 64 000.

If the value is unbounded and StorageClass has "iopsPerGB: 1000", then the user can create volumes up to 64GB and can't create 65GB volume, which is quite surprising. Kubernetes should cap the value at 64 000, so users can create volume of any size.

leakingtapan commented:
We can make this less surprising if proper error is surfaced. And that’s the behavior of EBS already. Don’t think this change is making it worse though

jsafrane commented:
It can makes PVC YAML files that were working broken, that's IMO quite serious.

philjb committed:
AWS: increase io1 volume IOPs limit

Max IOPS for SSD (io1) volumes was increased from 20,000 to 32,000:
https://aws.amazon.com/about-aws/whats-new/2017/12/amazon-ebs-provisioned-iops-ssd--io1--volumes-now-support-32-000-iops-and-500-mbs-per-volume/

and later to 64,000:
https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-elastic-block-store-announces-double-the-performance-of-provisioned-iops-volumes/

philjb commented:
On behalf of @jacobmarble - I changed this PR to just update the max IOPs limit for aws io1 volumes to 64,000 per AWS docs and @jsafrane's recommendation (also updated an error msg that was stale). 

Keeping with the simple change should prevent any existing configurations that specify a very fast IOPs setting from failing due to over requesting IOPs. (Note: anyone requesting more than 50 IOPs per GiB may get a failure from AWS as that is not checked and current existing k8s behavior.) 

@jsafrane @leakingtapan please review again and approve.

philjb commented:
/retest

leakingtapan commented:
/lgtm

jsafrane commented:
Thanks a lot!
/approve

jsafrane commented:
/retest

philjb commented:
@jsafrane - could you add the milestone?

jsafrane commented:
No. @saad-ali, can you please add 1.19 milestone? This one is harmless.

philjb commented:
@saad-ali - just following up here on adding the milestone. thanks for your help.

philjb commented:
@saad-ali - friendly weekly reminder here for the milestone.

philjb commented:
@jsafrane - could help find @saad-ali ?

saad-ali commented:
/assign

saad-ali commented:
/milestone 1.19

saad-ali commented:
/milestone v1.19

justaugustus commented:
/test pull-kubernetes-e2e-kind-ipv6
